sw/host/opentitanlib/src/proxy/errors.rs

```diff
@@ -0,0 +1,124 @@
+    fn as_anyhow_error(self: Box<Self>) -> anyhow::Error;
+}
+
+/// `impl_serializable_error` needs to be invoked for every error type
+/// that can be sent over the wire.
+#[macro_export]
+macro_rules! impl_serializable_error {
+    ($t:ty) => {
+        const _: () = {
+            use $crate::proxy::errors::SerializableError;
+            #[typetag::serde]
+            impl SerializableError for $t {
+                fn as_anyhow_error(self: Box<$t>) -> anyhow::Error {
+                    self.into()
+                }
+            }
+        };
+    };
+}
+
+/// `SerializedError` is the wire form of errors that can be sent through the proxy.
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+pub struct SerializedError {
+    pub description: String,
+    pub backtrace: String,
+    pub error: Option<Box<dyn SerializableError>>,
+}
+
+// Helper macro for building the `From` implementation that converts
+// native error types into their serialized form. This is a recursive
+// macro.
+macro_rules! to_serialized_error {
+    // This is the terminal state for this macro: only one type to convert.
+    // Parameters:
+    //   error - The error object.
+    //   msg - The error message (e.g. `to_string()`).
+    //   bt - The error backtrace.
+    //   box - A function which can box the error if needed.
+    //   t - The error type to convert.
+    ($error:expr, $msg:expr, $bt:expr, $box:expr, $t:ty $(,)?) => {{
+        match $error.downcast::<$t>() {
+            Ok(e) => return SerializedError {
+                description: $msg,
+                backtrace: $bt,
+                error: Some($box(e)),
+            },
+
+            Err(_) => return SerializedError {
+                description: $msg,
+                backtrace: $bt,
+                error: None,
+            },
+        }
+    }};
+
+    // This is the non-terminal state for this macro: many types to convert.
+    // Parameters:
+    //   error - The error object.
+    //   msg - The error message (e.g. `to_string()`).
+    //   bt - The error backtrace.
+    //   box - A function which can box the error if needed.
+    //   t:ts - The error types to convert.
+    ($error:expr, $msg:expr, $bt:expr, $box:expr, $t:ty, $($ts:ty),+ $(,)?) => {{
+        let e2 = match $error.downcast::<$t>() {
+            Ok(e) => return SerializedError {
+                description: $msg,
+                backtrace: $bt,
+                error: Some($box(e)),
+            },
+            Err(e) => e,
+        };
+        to_serialized_error!(e2, $msg, $bt, $box, $($ts,)*);
+    }};
+}
+
+/// Converts a `SerializedError` back into a real error type.
+impl From<SerializedError> for anyhow::Error {
+    fn from(ser: SerializedError) -> anyhow::Error {
+        let error = if let Some(error) = ser.error {
+            error.as_anyhow_error()
+        } else {
+            anyhow::Error::msg(ser.description)
+        };
+        if ser.backtrace == "<disabled>" {
```

Comment:
Should this be an Option instead?

Reply:
Maybe, but: `anyhow::Error::backtrace` always returns something, even if backtrace is disabled.  I could make `SerializableError::backtrace` and Option, but then I'd have to perform this check in the serialization rather than the deserialization.